### DISCUSSION TITLE
Update list of contents for multiple viewers

### DISCUSSION THREAD

**Discussion by manlioesp** [unpoly/unpoly#299] [General]
Hi to everybody, I need to update a table of contents during a live streaming event (the video is in an iframe from external server). The audience is about 400 people, and every 10/15 minutes it should be updated to show information about the new speaker and his contents.
I have never made a thing like this, and I thought to use Server Side Events to achieve this.
I think unpoly doesn't support this strategy, but is there an alternative way to do this with unpoly?
I don't want to overload the server so I need to find the right solution for the task.
Thank you very much.

**Comment by adam12** [comment]
I'd use [`[up-poll]`](https://unpoly.com/up-poll) for this. I don't think SSE is necessary here if you want to just do a reload every 10/15 minutes (you could even do something shorter if you wanted).

**Reply by manlioesp** [reply]
Thank you adam for your reply. I'm sorry for my bad explanation, but I meant that the speaker will change every 5/10/15 minutes but I don't know exactly when, so I thought that I had to poll every max 10s to have a decent sync between streaming and table of contents.
For this reason I was thinking about SSE.
Maybe I can use htmx together with unpoly?
Thak you again!

**Reply by adam12** [reply]
Well, you don't need HTMX for SSE. I didn't test this, but you could just make a compiler that handled this.

```javascript
up.compiler('[sse-reload]', (el) => {
  const url = el.getAttribute('sse-reload')
  const evtSource = new EventSource(url)

  evtSource.addEventListener('reload', () => {
    up.reload(el)
  })
  
  return () => {
    evtSource.close
  }
})
```

```html
<div sse-reload="/sse/speaker-change">
  Current speaker info
</div>
```

**Reply by manlioesp** [reply]
Thank you Adam! This is really useful, I'll try it as soon as possible.
Have a nice day.

**Comment by triskweline** [comment]
Wow, this implementation is really short.

I'd recommend to test this under load. Depending on your backend implementation of SSE this may open 400 threads to maintain client connections. It might perform better to have 400 people poll every 10 seconds and use [`[up-time]`](https://unpoly.com/up-time) to deliver empty responses if nothing changed.

I suppose this is less of an issue with event-based backends like Node.js.

**Comment by manlioesp** [comment]
Thank you Henning! Now users will be 1000, so the things are more complicated from the server point of view. I searched some information about this, and it seems that SSE should generally perform better against server polling, but your advice seems very interesting and from my point of view easier to implement thanks to Unpoly.
Do you think that 5 to 10 seconds polling with up-time could be a viable solution with 1000 users? unluckily I don't have experience with Node.js. 
At this time we have a dedicated VPS and I'm still trying to understand if I should use a different solution.

Anyway I was trying to make SSE working to give me an alternative route but I need to understand something more due to my inexperience.

I'm using Adam's code and my php server file is something like this
`header('Content-Type: text/event-stream');
echo date("h:i:sa");
flush();`

But it is not working and I can't understand exactly what I have to do.
Thank you for any advice.
Have a nice day.

**Reply by adam12** [reply]
> I'm using Adam's code and my php server file is something like this

I don't think that is sending a recognizable event (which in my example, I chose `reload`). Have you seen the PHP example for SSE? https://github.com/mdn/dom-examples/blob/master/server-sent-events/sse.php

Since you're using PHP, I'd probably follow Henning's suggestion of using a service like Pusher. But I'd probably take it a step further, since you're still going to be calling `up.reload` which might cause a thundering hurd against your server. I'd introduce random jitter for each client, so they receive the reload event, then have a random timeout before they make the request to the server.

**Reply by manlioesp** [reply]
Thank you, I didn't read this doc, but another on another website.
At the end I decided to use pusher with the help of my friend that is an expert JS developer.
It seems to work great.
Thank you for your support.

**Comment by triskweline** [comment]
> Do you think that 5 to 10 seconds polling with up-time could be a viable solution with 1000 users?

That really depends on the throughput of your app. If 1000 clients poll every 5 seconds you need to field at least 200 requests per second. This does not account for other jobs your server might have, like terminating HTTPS, requests to other pages or background jobs.

I must once more stress the important of testing your system under the load you expect. That means actually having 1000 clients connecting to your app. I remember how we once experienced a catastrophic failure for a system like this. A postmortem revealed that we had configured very modern SSL ciphers that were so expensive to compute that the load balancers ran into CPU limits, while the app servers were running idle.

> Anyway I was trying to make SSE working to give me an alternative route but I need to understand something more due to my inexperience.

There are two more, low-code alternatives to implement live updates for a large number of users:

- For SSE/WebSockets: Use an external service like [Pusher](https://pusher.com/) or [Ably](https://ably.com/) to maintain those 1000 connections for you. When you app has a new event, you can send it to those services who will in turn distribute it to your 1000 clients.
- For polling: Configure a web server to cache your polling response for 5 seconds. This way your (slow) app code will only run once every 5 seconds. You can configure caching in most HTTP servers, or use an external service like AWS CloudFront.

**Reply by manlioesp** [reply]
Thank you very much for your precious advices. The VPS is in charge of managing only some HTML and a simple form for asking question s.
Videoconference and everything else is managed by external services, so I will go also for this part with pusher or ably.
Thank you :-)
